data/openstreetmaps/src/main/java/de/mm20/launcher2/openstreetmaps/OsmSerialization.kt

```diff
@@ -37,7 +46,9 @@ class OsmLocationDeserializer : SearchableDeserializer {
             openingSchedule = null,
             websiteUrl = null,
             phoneNumber = null,
-            isCacheUpToDate = false,
+            updatedSelf = deferred {
+                osmRepository.get(id).firstOrNull()
+            }
```

Comment:
It is probably a good idea to remove the entry from the database if `.get(id)` returns `null`, since entries in OSM could also just as well be deleted. (I am not sure if IDs are re-used, but still.)

Just imagine a shop closing; That's enough of a reason for someone to delete that node in OSM.

Reply:
It could also mean that the user is offline, or that there was an error in the request

Reply:
I mean in case everything else works and the api returns 404. IIRC this is the behaviour for IDs that are not found.

Reply:
Something like this:

```kotlin
fun requestUpdatedDeferredSearchable() {
    ...
    viewModelScope.launch {
        this@SearchableItemVM.searchable.value = updated.await().fold(
            onSuccess = { v -> v },
            onFailure = { e -> if (e is HttpException) null else return@launch }
        )
    }
}
```

Reply:
> What about changing updatedSelf in `DeferredSearchable<T>` from `Deferred<T?>?` to `Deferred<Result<T>>?`?

Sounds good. But maybe we use a custom class instead. We need to differentiate between three different outcomes:
1. The data is retrieved successfully. The UI is updated with the new data. The new data should be fed back into the database
2. The data is temporarily unavailable (for example due to the user being offline). The UI should continue to show the old data, possibly with a hint that the result is cached (?)
3. The data is permanently unavailable. The data should be removed from the database.

> ```kt
>onFailure = { e -> if (e is HttpException) null else return@launch }
 >```

`HttpException` is an implementation detail specific to Retrofit. Not all providers use Retrofit internally (also I consider migrating to Ktor in the future, but that's a different story). Either we implement our own exception classes (`PermanentlyUnavailableException` and `TemporarilyUnavailableException`), or we use a different `sealed interface Result` with `Result.Success` `Result.PermanentlyUnavailable` and `Result.TemporarilyUnavailable`

Reply:
Okay, seems reasonable. But currently, how would we make sure that `SearchableItemVM` "unrenders" itself? `searchable` is of type `MutableStateFlow<SavableSearchable?>`, so we could assign `null` to its value. As far as I can see though every code accessing `value` defaults to some initial value passed, e.g. in `ListItem()`:
```kotlin
fun ListItem(item: SavableSearchable)
    ...
    val item = viewModel.searchable.collectAsState().value ?: item
```
(Maybe just change that to `:? return`?)

If this is unnecessary I guess we could just call `favoritesService.searchableRepository.delete(invalidLocation)` in `requestUpdatedDeferredSearchable()` and call it a day.

Anyway; If you are pressed on time I can come up with an implementation for this. My experience just tells me that two people working on the same branch can cause some headaches... 😄

Reply:
> Okay, seems reasonable. But currently, how would we make sure that SearchableItemVM "unrenders" itself?

They are stored in a cache with a max size of 500 and should be garbage collected when they aren't needed anymore.

> As far as I can see though every code accessing value defaults to some initial value passed, e.g. in `ListItem():`

Yeah, not sure about that. Basically, I want the searchable to come from the viewmodel so that we can update it from there, but the viewmodel is only populated in a `LaunchedEffect` in `ListItem`. Since side effects do not run synchronously, there is a chance that `viewModel.searchable` is `null` for a frame. So in that frame, we use the original parameter instead. There is probably a cleaner way to do this but I can't think of one right now.

> If this is unnecessary I guess we could just call favoritesService.searchableRepository.delete(invalidLocation) in requestUpdatedDeferredSearchable() and call it a day.

`searchableRepository` shouldn't be public, I don't know why it is, probably an oversight. But we could expose a new function that forwards that call to delete, yes. And another one that updates its contents

> Anyway; If you are pressed on time I can come up with an implementation for this. My experience just tells me that two people working on the same branch can cause some headaches... 😄

Sure, go ahead. There is plenty of other things I need to attend to before we can finally release this update.